core/iwasm/common/wasm_c_api.c

```diff
@@ -3385,9 +3423,13 @@ wasm_func_call(const wasm_func_t *func, const wasm_val_vec_t *params,
     if (argv != argv_buf)
         wasm_runtime_free(argv);
 
+#if WASM_ENABLE_DUMP_CALL_STACK != 0 && WASM_ENABLE_THREAD_MGR != 0
+    WASMCluster *cluster = wasm_exec_env_get_cluster(exec_env);
+    cluster_frames = cluster->exception_frames;
+#endif
     return wasm_trap_new_internal(
         func->store, func->inst_comm_rt,
-        wasm_runtime_get_exception(func->inst_comm_rt));
+        wasm_runtime_get_exception(func->inst_comm_rt), cluster_frames);
```

Comment:
@eloparco IIUC, here `func->inst_comm_rt` always points to the main module instance, so wasm_trap_new_internal uses the wrong frames, right? If we use the instance of the current thread, can the issue be resolved?
How about we adding an API in thread_mgr.c/h to lookup the exec_env of current thread, and then using `exec_env->module_inst` here instead of `func->inst_comm_rt`?
Somewhat like wasm_clusters_search_exec_env(module_inst), but traversing the clusters and finding the exec_env whose handle equals to os_self_thread().

Reply:
`wasm_trap_new_internal()` is called from the main thread and, at the time of the call, the exec_env of the spawned thread has already been freed here https://github.com/bytecodealliance/wasm-micro-runtime/blob/722381fcf2861596cb4c8ed7a6019deaabb14706/core/iwasm/libraries/thread-mgr/thread_manager.c#L668 so even saving in the cluster the thread id or directly the module instance of the crashed thread wouldn't work.

What about leaving it as it is? If a spawned thread crashes, the module instance reported in the frames is NULL. Not sure if there are use case in which the module instance stored in the frames is needed (not in mine at least).

Reply:
OK, a concern is that there may be data racing in accessing the `cluster->exception_frames`: suppose another thread is `wasm_cluster_set_exception`. Had better acquire the `cluster->lock` with `wasm_cluster_traverse_lock` before calling `wasm_trap_new_internal`?